packages/drawer/index.tsx

```diff
@@ -0,0 +1,253 @@
+    className: '',
+    children: null,
+    open: false,
+    onOpen: () => {},
+    onClose: () => {},
+    onTransitionEnd: () => {},
+    onKeyDown: () => {},
+    tag: 'aside',
+    dismissible: false,
+    modal: false,
+  };
+
+  componentDidMount() {
+    const {open} = this.props;
+    this.initFoundation();
+    if (open && this.foundation_) {
+      this.foundation_.open();
+    }
+  }
+
+  initFoundation = () => {
+    const {dismissible, modal} = this.props;
+    if (this.foundation_) {
+      this.foundation_.destroy();
+    }
+    if (dismissible) {
+      this.foundation_ = new MDCDismissibleDrawerFoundation(this.adapter);
+      this.foundation_.init();
+    } else if (modal) {
+      this.initializeFocusTrap();
+      this.foundation_ = new MDCModalDrawerFoundation(this.adapter);
+      this.foundation_.init();
+    }
+  };
+
+  componentDidUpdate(prevProps: DrawerProps & React.HTMLProps<HTMLElement>) {
+    const {dismissible, modal, open} = this.props;
+    const changedToModal = prevProps.modal !== this.props.modal;
+    const changedToDismissible =
+      prevProps.dismissible !== this.props.dismissible;
+    if (!dismissible && !modal) return;
+    if (changedToModal || changedToDismissible) {
+      this.initFoundation();
+    }
+    if (open !== prevProps.open) {
+      open ? this.foundation_.open() : this.foundation_.close();
+    }
+  }
+
+  componentWillUnmount() {
+    if (!this.foundation_) return;
+    this.foundation_.destroy();
+  }
+
+  initializeFocusTrap = () => {
+    this.focusTrap_ = util.createFocusTrapInstance(this.drawerElement_.current);
+  };
+
+  get classes() {
+    const {classList} = this.state;
+    const {className, dismissible, modal} = this.props;
+    return classnames('mdc-drawer', Array.from(classList), className, {
+      'mdc-drawer--dismissible': dismissible,
+      'mdc-drawer--modal': modal,
+    });
+  }
+
+  get adapter() {
+    return {
+      addClass: (className: string) => {
+        const {classList} = this.state;
+        classList.add(className);
+        this.setState({classList});
+      },
+      removeClass: (className: string) => {
+        const {classList} = this.state;
+        classList.delete(className);
+        this.setState({classList});
+      },
+      hasClass: (className: string) => this.classes.split(' ').includes(className),
+      elementHasClass: (element: HTMLElement, className: string) =>
+        element.classList.contains(className),
+      saveFocus: () => {
+        this.previousFocus_ = document.activeElement as HTMLElement | null;
+      },
+      restoreFocus: () => {
+        const hasPreviousFocus =
+          this.previousFocus_ && this.previousFocus_.focus;
+        const drawerElement =
+          this.drawerElement_ && this.drawerElement_.current;
+        if (
+          drawerElement &&
+          hasPreviousFocus &&
+          drawerElement.contains(document.activeElement)
+        ) {
+          this.previousFocus_!.focus();
```

Comment:
Is the ! here intentional? Formatting of this if statement also seems a bit weird.

Reply:
UGH! it was the formatter...
And yes is intentional. For some reason the tslint doesn't see the IF check I have for this.previousFocus